kolibri/core/content/utils/content_request.py

```diff
@@ -0,0 +1,525 @@
+    Lessons and Exams. Any model that attaches the `ContentAssignmentManager` will allow this.
+
+    :param dataset_id: The UUID of the synced dataset
+    :param transfer_session: The sync's transfer session model
+    :type transfer_session: morango.models.core.TransferSession
+    """
+    facility = Facility.objects.get(dataset_id=dataset_id)
+
+    new_contentnode_ids = set()
+
+    # process the new assignments
+    logger.info("Processing new content assignment requests")
+    for assignment in ContentAssignmentManager.find_all_downloadable_assignments(
+        transfer_session.id
+    ):
+        related_removals = ContentRemovalRequest.objects.filter(
+            reason=ContentRequestReason.SyncInitiated,
+            source_model=assignment.source_model,
+            source_id=assignment.source_id,
+        )
+        # delete any related removal requests
+        related_removals.delete()
+
+        ContentDownloadRequest.objects.get_or_create(
+            defaults=dict(
+                facility_id=facility.id,
+                reason=ContentRequestReason.SyncInitiated,
+                status=ContentRequestStatus.Pending,
+            ),
+            source_model=assignment.source_model,
+            source_id=assignment.source_id,
+            contentnode_id=assignment.contentnode_id,
+        )
+        new_contentnode_ids.update([assignment.contentnode_id])
+
+    # process new removals
+    logger.info("Processing new content removal requests")
+    for assignment in ContentAssignmentManager.find_all_removable_assignments(
+        transfer_session.id
+    ):
+        related_downloads = ContentDownloadRequest.objects.filter(
+            reason=ContentRequestReason.SyncInitiated,
+            source_model=assignment.source_model,
+            source_id=assignment.source_id,
+        )
+
+        if isinstance(assignment, DeletedAssignment):
+            # for completed downloads, we'll go through contentnode_ids and add removals
+            removed_contentnode_ids = related_downloads.values_list(
+                "contentnode_id", flat=True
+            ).distinct()
+        else:
+            removed_contentnode_ids = [assignment.contentnode_id]
+
+        for contentnode_id in removed_contentnode_ids:
+            ContentRemovalRequest.objects.get_or_create(
+                defaults=dict(
+                    facility_id=facility.id,
+                    reason=ContentRequestReason.SyncInitiated,
+                    status=ContentRequestStatus.Pending,
+                ),
+                source_model=assignment.source_model,
+                source_id=assignment.source_id,
+                contentnode_id=contentnode_id,
+            )
+
+        # delete any related download requests
+        related_downloads.delete()
+
+    # avoid adding extra logs
+    if not new_contentnode_ids:
+        return
+
+    # TODO: we can't yet do this during a sync because the sync command creates a lock / transaction
```

Comment:
Hrm... odd, the db_lock doesn't seem to do anything special to make sure the transaction is shared between Django and SQLAlchemy: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/utils/lock.py#L37

Not sure what is different about the transaction created during the sync - it should be fine to created a nested transaction within that parent transaction.

Reply:
Does the SQLAlchemy side use a separate DB connection? That was my assumption for why it was blocking, for the reason that sqlite uses serializable transactions, so only one connection's transaction can execute at a time. If it does use the same connection, or if we can make it do that, my guess is that is shouldn't block like you said.

Reply:
Yes, it should be using the same connection, precisely to allow the db lock to be used - we use a special connection generator for SQLAlchemy that references the Django connection.

Reply:
So this change seems to have fixed it. It appears it wasn't sharing the connection after all, but looking at tests, this seemed like the appropriate change
https://github.com/learningequality/kolibri/pull/10365/commits/1baed22525776e5353ab5fde7ad78187e84f47fa#diff-76f6082e35fb73d50fa01980aa366b8455b9c2488734ab0d985582ab11a584bcR110